src/allmydata/test/common_util.py

```diff
@@ -78,13 +89,114 @@ def flip_one_bit(s, offset=0, size=None):
     return result
 
 
+class ReallyEqualMixin(object):
+    def failUnlessReallyEqual(self, a, b, msg=None):
+        self.assertEqual(a, b, msg)
+        self.assertEqual(type(a), type(b), "a :: %r (%s), b :: %r (%s), %r" % (a, type(a), b, type(b), msg))
+
+
+class SignalMixin(object):
+    # This class is necessary for any code which wants to use Processes
+    # outside the usual reactor.run() environment. It is copied from
+    # Twisted's twisted.test.test_process . Note that Twisted-8.2.0 uses
+    # something rather different.
+    sigchldHandler = None
+
+    def setUp(self):
+        # make sure SIGCHLD handler is installed, as it should be on
+        # reactor.run(). problem is reactor may not have been run when this
+        # test runs.
+        if hasattr(reactor, "_handleSigchld") and hasattr(signal, "SIGCHLD"):
+            self.sigchldHandler = signal.signal(signal.SIGCHLD,
+                                                reactor._handleSigchld)
+        return super(SignalMixin, self).setUp()
+
+    def tearDown(self):
+        if self.sigchldHandler:
+            signal.signal(signal.SIGCHLD, self.sigchldHandler)
+        return super(SignalMixin, self).tearDown()
+
+
 class StallMixin(object):
     def stall(self, res=None, delay=1):
         d = defer.Deferred()
         reactor.callLater(delay, d.callback, res)
         return d
 
 
+class Marker(object):
+    pass
+
+class FakeCanary(object):
+    """For use in storage tests.
+
+    Can be moved back to test_storage.py once enough Python 3 porting has been
```

Comment:
Maybe move this to test_storage.py? Unless it's used in multiple storage test modules.

Reply:
I am seeing it used in several other test modules:

```
test_storage_web.py
53:from .common_util import FakeCanary
310:        canary = FakeCanary()

web/test_introducer.py
37:    FakeCanary,
197:                FakeCanary(),
208:            FakeCanary(),

common_util.py
130:class FakeCanary(object):

test_storage.py
55:from .common_util import FakeCanary
127:                          FakeCanary())
137:                          FakeCanary())
231:                          FakeCanary())
390:            canary = FakeCanary()
547:        canary = FakeCanary()
580:        canary = FakeCanary(True)
652:        canary = FakeCanary()

test_crawler.py
30:from allmydata.test.common_util import StallMixin, FakeCanary
130:                                              set([0]), 99, FakeCanary())
```

Reply:
But in that case ... https://github.com/tahoe-lafs/tahoe-lafs/pull/846.